cpp/src/groupby/sort/group_collect.cu

```diff
@@ -17,30 +17,96 @@
+  };
+
+  // Purge null entries in grouped values.
+  auto null_purged_entries =
+    cudf::detail::copy_if(table_view{{values}}, not_null_pred, stream, mr)->release();
+
+  std::unique_ptr<column> &null_purged_values = null_purged_entries[0];
+
+  // Recalculate offsets after null entries are purged.
+  auto null_purged_sizes = make_numeric_column(
+    data_type{type_to_id<size_type>()}, num_groups, mask_state::UNALLOCATED, stream, mr);
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(num_groups),
+    null_purged_sizes->mutable_view().template begin<size_type>(),
+    [d_offsets = offsets.template begin<size_type>(), not_null_pred] __device__(auto i) {
+      return thrust::count_if(thrust::seq,
+                              thrust::make_counting_iterator<size_type>(d_offsets[i]),
+                              thrust::make_counting_iterator<size_type>(d_offsets[i + 1]),
+                              not_null_pred);
+    });
+
+  auto null_purged_offsets = strings::detail::make_offsets_child_column(
+    null_purged_sizes->view().template begin<size_type>(),
+    null_purged_sizes->view().template end<size_type>(),
+    stream,
+    mr);
+
+  return std::make_pair<std::unique_ptr<column>, std::unique_ptr<column>>(
+    std::move(null_purged_values), std::move(null_purged_offsets));
+}
+
 std::unique_ptr<column> group_collect(column_view const &values,
                                       cudf::device_span<size_type const> group_offsets,
                                       size_type num_groups,
+                                      null_policy null_handling,
                                       rmm::cuda_stream_view stream,
                                       rmm::mr::device_memory_resource *mr)
 {
-  rmm::device_buffer offsets_data(
-    group_offsets.data(), group_offsets.size() * sizeof(cudf::size_type), stream, mr);
+  auto offsets_column = make_numeric_column(
+    data_type(type_to_id<size_type>()), num_groups + 1, mask_state::UNALLOCATED, stream, mr);
+
+  thrust::copy(rmm::exec_policy(stream),
+               group_offsets.begin(),
+               group_offsets.end(),
+               offsets_column->mutable_view().template begin<size_type>());
+
+  std::unique_ptr<cudf::column> child_column;
 
-  auto offsets = std::make_unique<cudf::column>(
-    cudf::data_type(cudf::type_to_id<cudf::size_type>()), num_groups + 1, std::move(offsets_data));
+  // If column of grouped values contains null elements, and null_policy == EXCLUDE,
+  // those elements must be filtered out, and offsets recomputed.
+  if (null_handling == null_policy::EXCLUDE && values.has_nulls()) {
+    std::tie(child_column, offsets_column) = cudf::groupby::detail::purge_null_entries(
+      values, offsets_column->view(), num_groups, stream, mr);
+  } else {
+    child_column = std::make_unique<cudf::column>(values, stream, mr);
+  }
```

Comment:
This is bad. Declare and assign later is discouraged. Please use a lambda with instant execution. For examle:
```
auto [child_column, offsets_column] = [null_handling, num_groups, &values, &group_offsets, stream, mr]{
    if (null_handling == null_policy::EXCLUDE && values.has_nulls()) {
     return cudf::groupby::detail::purge_null_entries(
      values, offsets_column->view(), num_groups, stream, mr);
  } else {
    auto child_column = std::make_unique<cudf::column>(values, stream, mr);
    auto offsets_column = make_numeric_column(
        data_type(type_to_id<size_type>()), num_groups + 1, mask_state::UNALLOCATED, stream, mr);

  thrust::copy(rmm::exec_policy(stream),
               group_offsets.begin(),
               group_offsets.end(),
               offsets_column->mutable_view().template begin<size_type>());
    return { child_column, offsets_column};
  }
} ();
```

Reply:
Now that we are on C++17, I should do this in my reference code as well, when I touch it next.

Reply:
I've reworked this part with a lambda.